File: apps/android/app/src/main/kotlin/app/trackevolution/ui/charts/HealthStrip.kt
package app.trackevolution.ui.charts

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.testTag
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import app.trackevolution.core.Health
import app.trackevolution.core.Units
import app.trackevolution.core.model.SessionChannels
import app.trackevolution.ui.LayoutClass
import app.trackevolution.ui.LocalLayoutMetrics
import app.trackevolution.ui.LocalUnitSystem
import app.trackevolution.ui.theme.TrackCard
import app.trackevolution.ui.theme.TrackTheme
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

/**
 * Every figure shows in the account's system — °F and psi for an imperial
 * account, °C and kPa for a metric one — while the maths stays in the stored
 * units. Read from `LocalUnitSystem`, so it is a composable getter: capture it
 * into a local before handing it to anything that isn't.
 */
private val UNITS: Health.Units
    @Composable get() = Units.healthUnits(LocalUnitSystem.current)

/**
 * The session health strip (#190) — the counterpart of `healthStripHtml` and
 * `sparklineSvg` in `public/js/health.js`.
 *
 * The panel's Car tab: one card per figure the import stored — peak oil,
 * coolant and transmission temperature, minimum oil pressure and battery, fuel
 * and the four tyre pressures at lap end, peak tyre temperature per corner, and
 * per-lap peak boost — carrying the session's number by that channel's own
 * rule, a sparkline across the laps with its threshold bands shaded, and the
 * cross-corner tyre spread and fuel outlook under them. The maths is `Health`
 * in `:core`, pinned to the web implementation by `contracts/logic/health.json`;
 * this file draws.
 *
 * Four things are load-bearing.
 *
 * **Thresholds shade, they never alarm.** A card past its watch line takes a
 * tinted border and one past its `over` line a tinted background, in the
 * garage's own wear colours — `danger` — rather than a second severity scale
 * invented here. Nothing blocks, nothing pops.
 *
 * **The number is the importer's reduction, and the card says which** ("peak",
 * "min", "at lap end"), because "oil 134 °C" means nothing without knowing it is
 * the lap's peak rather than its average.
 *
 * **The figures show in the account's system** ([UNITS]) while the maths stays
 * in the stored units — the conversion is the last step, exactly as on the web.
 *
 * **The web's per-lap table is deliberately absent.** Fifteen columns is a desk
 * layout; on a phone the sparkline carries the shape and the highlighted laps
 * are marked on it in their slot colours, which is the same question answered in
 * the space available.
 */
@Composable
fun HealthStrip(
    channels: SessionChannels,
    /** Channel-lap indexes in slot order, as [LapChannelChart] keeps them. */
    lit: List<Int>,
    slots: List<Color>,
    modifier: Modifier = Modifier,
    /** The session's own lap number for a channel entry, for the per-lap table. */
    lapNumber: (Int) -> Int = { it + 1 },
) {
    val sh = remember(channels) { Health.sessionHealth(channels) } ?: return
    val order = sh.laps.map { it.chIdx }
    val colors = TrackTheme.colors
    // Two on a phone; as many as the column holds above that. 150dp is the web's
    // own `.health-cards` minimum.
    val perRow = maxOf(2, LocalLayoutMetrics.current.columns(150.dp, gap = 8.dp))

    val summary = Health.healthSummary(channels, UNITS)
        ?: sh.columns.mapNotNull { column ->
            Health.defFor(column.key)?.let {
                "${it.label} ${Health.displayValue(it, column.extreme.v, UNITS).text}"
            }
        }.joinToString(", ")

    Column(modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(10.dp)) {
    TrackCard(
        Modifier
            .fillMaxWidth()
            .clearAndSetSemantics {
                testTag = "healthStrip"
                contentDescription = "Car: the session's health figures. $summary"
            },
    ) {
        Column(Modifier.padding(12.dp), verticalArrangement = Arrangement.spacedBy(10.dp)) {
            Text("Car", style = TrackTheme.typography.xxs, color = colors.textMuted)
            Text(
                "What the car was doing while you drove it. One figure per lap, reduced the way " +
                    "the recorder stored it.",
                style = TrackTheme.typography.xxs,
                color = colors.textFaint,
            )
            for ((groupKey, groupLabel) in Health.HEALTH_GROUPS) {
                val columns = sh.columns.filter { Health.defFor(it.key)?.group == groupKey }
                if (columns.isEmpty()) continue
                Text(groupLabel, style = TrackTheme.typography.xxs, color = colors.textFaint)
                // Two to a row on a phone, which is what fits legibly; more once
                // the column is wide enough for them (NS-34 ticket 3). Still not
                // a LazyVGrid — one inside an already-scrolling panel nests
                // scrolls, which is why this chunks by hand.
                columns.chunked(perRow).forEach { group ->
                    Row(
                        Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.spacedBy(8.dp),
                    ) {
                        group.forEach { column ->
                            FigureCard(
                                column = column,
                                order = order,
                                lit = lit,
                                slots = slots,
                                modifier = Modifier.weight(1f),
                            )
                        }
                        repeat(perRow - group.size) { Spacer(Modifier.weight(1f)) }
                    }
                }
            }
            SpreadLines(channels)
            FuelLine(channels)
        }
    }

    // The per-lap table (NS-34 ticket 3), at expanded width only: the cards
    // above reduce each column to one number precisely because fifteen columns
    // do not fit a phone, and this is that reduction undone where there is room.
    //
    // Its **own card**, deliberately, rather than the last thing inside the one
    // above. That card declares itself a single accessibility element — a
    // fifteen-by-N grid of numbers is unbearable to swipe through one cell at a
    // time — and anything inside it inherits that, so a table placed there would
    // be invisible to TalkBack as well as to a test. As a sibling it carries a
    // summary of its own.
    if (LocalLayoutMetrics.current.layoutClass == LayoutClass.Expanded) {
        PerLapTable(sh = sh, lit = lit, slots = slots, lapNumber = lapNumber)
    }
    }
}

@Composable
private fun FigureCard(
    column: Health.Column,
    order: List<Int>,
    lit: List<Int>,
    slots: List<Color>,
    modifier: Modifier = Modifier,
) {
    val def = Health.defFor(column.key) ?: return
    val colors = TrackTheme.colors
    val display = Health.displayValue(def, column.extreme.v, UNITS)
    val border = when (column.status) {
        Health.Status.DUE -> colors.danger
        Health.Status.LOW -> colors.danger.copy(alpha = 0.45f)
        else -> colors.borderHairline
    }
    val background = if (column.status == Health.Status.DUE) colors.dangerTint else Color.Transparent
    val statusWord = when (column.status) {
        Health.Status.DUE -> "over the line"
        Health.Status.LOW -> "watch"
        else -> null
    }
    val line = colors.chartLine

    Column(
        modifier
            .background(background, RoundedCornerShape(10.dp))
            .border(1.dp, border, RoundedCornerShape(10.dp))
            .padding(10.dp),
        verticalArrangement = Arrangement.spacedBy(2.dp),
    ) {
        Text(
            def.label,
            style = TrackTheme.typography.xs,
            color = colors.textMuted,
            maxLines = 1,
        )
        Row(verticalAlignment = Alignment.Bottom, horizontalArrangement = Arrangement.spacedBy(4.dp)) {
            Text(
                display.text,
                style = TrackTheme.typography.h3,
                color = colors.textStrong,
                maxLines = 1,
            )
            statusWord?.let {
                Text(it, style = TrackTheme.typography.xxs, color = colors.dangerInk, maxLines = 1)
            }
        }
        Text(ruleWord(def), style = TrackTheme.typography.xxs, color = colors.textFaint)
        val danger = colors.danger
        Canvas(Modifier.fillMaxWidth().height(34.dp)) {
            drawSparkline(def, column, order, lit, slots, line, danger)
        }
    }
}

/** The importer's rule, said out loud — `ruleWord` in the JS. */
private fun ruleWord(def: Health.Def): String = when (def.reduce) {
    Health.Reduce.MAX -> "peak"
    Health.Reduce.MIN -> "min"
    Health.Reduce.END -> "at lap end"
}

/**
 * One column across the session's health laps: x is the lap's position among
 * them (so every card lines up), the threshold bands shaded behind the line, and
 * the highlighted laps marked in their slot colours.
 */
private fun DrawScope.drawSparkline(
    def: Health.Def,
    column: Health.Column,
    order: List<Int>,
    lit: List<Int>,
    slots: List<Color>,
    lineColor: Color,
    danger: Color,
) {
    val series = column.series
    if (series.isEmpty() || size.width <= 0f || size.height <= 0f) return
    val pad = 4f * density
    val n = max(2, order.size)
    val index = order.withIndex().associate { (i, chIdx) -> chIdx to i }

    var y0 = series.minOf { it.v }
    var y1 = series.maxOf { it.v }
    // Bring the lines into view when the data sits near them, so the shading
    // says how close the session came rather than only whether it crossed.
    val watch = def.watch
    val over = def.over
    if (watch != null && over != null) {
        val near = if (def.low) watch * 1.05 else watch * 0.95
        if (if (def.low) y0 < near else y1 > near) {
            y0 = min(y0, over)
            y1 = max(y1, over)
        }
    }
    val ypad = max((y1 - y0) * 0.15, 1e-6)
    y0 -= ypad
    y1 += ypad

    fun x(i: Int) = pad + i.toFloat() / (n - 1) * (size.width - pad * 2)
    fun y(v: Double) = pad + ((y1 - v) / (y1 - y0)).toFloat() * (size.height - pad * 2)
    fun clampY(v: Double) = min(max(y(v), pad), size.height - pad)

    // The bands: past `over` in the danger tint, between the lines lighter.
    if (watch != null && over != null) {
        fun band(from: Double, to: Double, alpha: Float) {
            val ya = clampY(to)
            val yb = clampY(from)
            if (abs(yb - ya) <= 0.5f) return
            drawRect(
                color = danger.copy(alpha = alpha),
                topLeft = Offset(0f, min(ya, yb)),
                size = Size(size.width, abs(yb - ya)),
            )
        }
        band(if (def.low) y0 else over, if (def.low) over else y1, 0.18f)
        band(if (def.low) over else watch, if (def.low) watch else over, 0.09f)
    }

    val path = Path()
    series.forEachIndexed { i, s ->
        val px = x(index[s.chIdx] ?: i)
        val py = y(s.v)
        if (i == 0) path.moveTo(px, py) else path.lineTo(px, py)
    }
    drawPath(path, color = lineColor, style = Stroke(width = 1.5f * density))

    // The highlighted laps, in the slot colours the chips use.
    lit.forEachIndexed { slot, chIdx ->
        val s = series.firstOrNull { it.chIdx == chIdx } ?: return@forEachIndexed
        val i = index[chIdx] ?: return@forEachIndexed
        drawCircle(
            color = slots[slot % slots.size],
            radius = 2.5f * density,
            center = Offset(x(i), y(s.v)),
        )
    }
}

/**
 * Cross-corner spread is the figure a setup change is judged by, so it gets
 * words rather than a chart: left−right on each axle and front−rear, for the
 * last lap that carried all four corners.
 */
@Composable
private fun SpreadLines(channels: SessionChannels) {
    val temps = remember(channels) { Health.sessionSpread(channels, "tyreC").lastOrNull() }
    val pressures = remember(channels) { Health.sessionSpread(channels, "tyreKpa").lastOrNull() }
    if (temps == null && pressures == null) return
    val colors = TrackTheme.colors
    val units = UNITS

    fun text(def: Health.Def, spread: Health.LapSpread, label: String): String {
        val front = Health.displayDelta(def, spread.front, units).text
        val rear = Health.displayDelta(def, spread.rear, units).text
        val axle = Health.displayDelta(def, spread.axle, units).text
        return "$label — front L−R $front, rear L−R $rear, front−rear $axle"
    }

    Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
        Text("Cross-corner spread", style = TrackTheme.typography.xxs, color = colors.textFaint)
        temps?.let { spread ->
            Health.defFor("tyreCLF")?.let {
                Text(
                    text(it, spread, "Tyre temps"),
                    style = TrackTheme.typography.xs,
                    color = colors.textMuted,
                )
            }
        }
        pressures?.let { spread ->
            Health.defFor("tyreKpaLF")?.let {
                Text(
                    text(it, spread, "Tyre pressures"),
                    style = TrackTheme.typography.xs,
                    color = colors.textMuted,
                )
            }
        }
    }
}

@Composable
private fun FuelLine(channels: SessionChannels) {
    val fuel = remember(channels) { Health.fuelBurn(channels) } ?: return
    Text(
        "Fuel — ${fuel.perLapPct.roundToInt()}% a lap, ${fuel.lastPct.roundToInt()}% left: " +
            "≈${fuel.lapsRemaining} lap${if (fuel.lapsRemaining == 1) "" else "s"} at this rate",
        style = TrackTheme.typography.xs,
        color = TrackTheme.colors.textMuted,
    )
}

/**
 * Every figure, every lap — the port of `healthTableHtml` (NS-34 ticket 3).
 *
 * Columns are `sh.columns`: `HEALTH_DEFS` order, filtered to what the session
 * actually stored, so the table and the cards above can never disagree about
 * which figures exist. It scrolls horizontally rather than squeezing, because a
 * temperature narrowed to three digits is not a temperature any more.
 */
@Composable
private fun PerLapTable(
    sh: Health.SessionHealth,
    lit: List<Int>,
    slots: List<Color>,
    lapNumber: (Int) -> Int,
) {
    if (sh.laps.isEmpty() || sh.columns.isEmpty()) return
    val colors = TrackTheme.colors
    TrackCard(Modifier.fillMaxWidth()) {
    Column(Modifier.padding(12.dp), verticalArrangement = Arrangement.spacedBy(6.dp)) {
    Text("Per lap", style = TrackTheme.typography.xxs, color = colors.textFaint)
    Column(
        Modifier.horizontalScroll(rememberScrollState()),
        verticalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
            Text(
                "Lap",
                style = TrackTheme.typography.xxs,
                color = colors.textFaint,
                modifier = Modifier.width(72.dp),
            )
            sh.columns.forEach { column ->
                val def = Health.defFor(column.key)
                Column(Modifier.width(64.dp), horizontalAlignment = Alignment.End) {
                    Text(
                        def?.label ?: column.key,
                        style = TrackTheme.typography.xxs,
                        color = colors.textFaint,
                        maxLines = 1,
                    )
                    if (def != null) {
                        Text(
                            Health.displayValue(def, 0.0, UNITS).unit,
                            style = TrackTheme.typography.xxs,
                            color = colors.textFaint,
                        )
                    }
                }
            }
        }
        sh.laps.forEach { row ->
            Row(
                horizontalArrangement = Arrangement.spacedBy(12.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Row(
                    Modifier.width(72.dp),
                    horizontalArrangement = Arrangement.spacedBy(5.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    // The slot colour when this lap is lit, so a row lines up
                    // with the traces above — never colour alone, since the lap
                    // number is right beside it.
                    val slot = lit.indexOf(row.chIdx)
                    if (slot >= 0 && slot < slots.size) {
                        Box(Modifier.size(7.dp).clip(CircleShape).background(slots[slot]))
                    }
                    Text(
                        "Lap ${lapNumber(row.chIdx)}",
                        style = TrackTheme.typography.xs,
                        color = colors.textStrong,
                        maxLines = 1,
                    )
                }
                sh.columns.forEach { column ->
                    val def = Health.defFor(column.key)
                    val value = row.values[column.key]
                    Text(
                        if (def != null && value != null) {
                            Health.displayValue(def, value, UNITS).text
                        } else {
                            "—"
                        },
                        style = TrackTheme.typography.xs,
                        color = if (value == null) colors.textFaint else colors.textStrong,
                        maxLines = 1,
                        modifier = Modifier.width(64.dp),
                        textAlign = TextAlign.End,
                    )
                }
            }
        }
    }
    }
    }
}

File: apps/ios/App/Charts/HealthStrip.swift
import SwiftUI
import TrackEvolutionKit

/// The session health strip (#190) — the counterpart of `healthStripHtml` and
/// `sparklineSvg` in `public/js/health.js`.
///
/// The panel's Car tab: one card per figure the import stored — peak oil,
/// coolant and transmission temperature, minimum oil pressure and battery, fuel
/// and the four tyre pressures at lap end, peak tyre temperature per corner,
/// and per-lap peak boost — carrying the session's number by that channel's own
/// rule, a sparkline across the laps with its threshold bands shaded, and the
/// cross-corner tyre spread and fuel outlook under them. The maths is the Kit's
/// `Health`, pinned to the web implementation by `contracts/logic/health.json`;
/// this file draws.
///
/// Four things are load-bearing.
///
/// **Thresholds shade, they never alarm.** A card past its watch line takes a
/// tinted border and one past its `over` line a tinted background, in the
/// garage's own wear colours — `danger` — rather than a second severity scale
/// invented here. Nothing blocks, nothing pops.
///
/// **The number is the importer's reduction, and the card says which** ("peak",
/// "min", "at lap end"), because "oil 134 °C" means nothing without knowing it
/// is the lap's peak rather than its average.
///
/// **The figures show in °F and psi** (`Health.Units.us`), matching the rest of
/// the logbook, while the maths stays in the stored units — the conversion is
/// the last step, exactly as on the web.
///
/// **The web's per-lap table is deliberately absent.** Fifteen columns is a
/// desk layout; on a phone the sparkline carries the shape and the highlighted
/// laps are marked on it in their slot colours, which is the same question
/// answered in the space available.
struct HealthStrip: View {
    let channels: SessionChannels
    /// Channel-lap indexes in slot order, as `LapChannelPanel` keeps them.
    let lit: [Int]
    let slots: [Color]
    /// The session's own lap number for a channel entry.
    let lapNumber: (Int) -> Int

    private static let units = Health.Units.us

    var body: some View {
        if let sh = Health.sessionHealth(channels) {
            let order = sh.laps.map(\.chIdx)
            TECard {
                VStack(alignment: .leading, spacing: 12) {
                    VStack(alignment: .leading, spacing: 2) {
                        Text("Car")
                            .teStyle(.eyebrow)
                            .foregroundStyle(Color(.textMuted))
                        Text(
                            "What the car was doing while you drove it. One figure per lap, "
                                + "reduced the way the recorder stored it."
                        )
                        .teStyle(.xxs)
                        .foregroundStyle(Color(.textFaint))
                    }
                    ForEach(Health.HEALTH_GROUPS.map(\.key), id: \.self) { group in
                        let columns = sh.columns.filter { Health.defFor($0.key)?.group == group }
                        if !columns.isEmpty {
                            VStack(alignment: .leading, spacing: 6) {
                                Text(Health.HEALTH_GROUPS.first { $0.key == group }?.label ?? group)
                                    .teStyle(.eyebrow)
                                    .foregroundStyle(Color(.textFaint))
                                LazyVGrid(
                                    columns: [GridItem(.adaptive(minimum: 140), spacing: 8)], spacing: 8
                                ) {
                                    ForEach(columns, id: \.key) { column in
                                        card(column, order: order)
                                    }
                                }
                            }
                        }
                    }
                    spreadLines()
                    fuelLine()
                }
            }
            .accessibilityElement(children: .ignore)
            .accessibilityLabel("Car: the session's health figures")
            .accessibilityValue(summary(sh))
            .accessibilityIdentifier("healthStrip")
        }
    }

    // MARK: - One figure

    @ViewBuilder
    private func card(_ column: Health.Column, order: [Int]) -> some View {
        if let def = Health.defFor(column.key) {
            let display = Health.displayValue(def, column.extreme.v, Self.units)
            VStack(alignment: .leading, spacing: 2) {
                Text(def.label)
                    .teStyle(.xs)
                    .foregroundStyle(Color(.textMuted))
                    .lineLimit(1)
                HStack(alignment: .firstTextBaseline, spacing: 4) {
                    Text(display.text)
                        .teStyle(.h3)
                        .monospacedDigit()
                        .foregroundStyle(Color(.textStrong))
                        .lineLimit(1)
                        .minimumScaleFactor(0.7)
                    if let word = statusWord(column.status) {
                        Text(word)
                            .teStyle(.xxs)
                            .foregroundStyle(Color(.dangerInk))
                    }
                }
                Text(ruleWord(def))
                    .teStyle(.xxs)
                    .foregroundStyle(Color(.textFaint))
                Canvas { context, size in
                    sparkline(def, column, order: order, in: context, size: size)
                }
                .frame(height: 34)
                .padding(.top, 2)
            }
            .padding(10)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(column.status == .due ? Color(.dangerTint) : Color.clear, in: .rect(cornerRadius: 10))
            .overlay(
                RoundedRectangle(cornerRadius: 10)
                    .stroke(borderColor(column.status), lineWidth: 1)
            )
        }
    }

    /// The importer's rule, said out loud — `ruleWord` in the JS.
    private func ruleWord(_ def: Health.Def) -> String {
        switch def.reduce {
        case .max: "peak"
        case .min: "min"
        case .end: "at lap end"
        }
    }

    private func statusWord(_ status: Health.Status?) -> String? {
        switch status {
        case .due: "over the line"
        case .low: "watch"
        default: nil
        }
    }

    private func borderColor(_ status: Health.Status?) -> Color {
        switch status {
        case .due: Color(.danger)
        case .low: Color(.danger).opacity(0.45)
        default: Color(.borderHairline)
        }
    }

    // MARK: - Sparkline

    /// One column across the session's health laps: x is the lap's position
    /// among them (so every card lines up), the threshold bands shaded behind
    /// the line, and the highlighted laps marked in their slot colours.
    private func sparkline(
        _ def: Health.Def, _ column: Health.Column, order: [Int],
        in context: GraphicsContext, size: CGSize
    ) {
        let series = column.series
        guard !series.isEmpty, size.width > 0, size.height > 0 else { return }
        let pad: CGFloat = 4
        let n = max(2, order.count)
        var index: [Int: Int] = [:]
        for (i, chIdx) in order.enumerated() { index[chIdx] = i }

        var y0 = series.map(\.v).min() ?? 0
        var y1 = series.map(\.v).max() ?? 1
        // Bring the lines into view when the data sits near them, so the shading
        // says how close the session came rather than only whether it crossed.
        if let watch = def.watch, let over = def.over {
            let near = def.low ? watch * 1.05 : watch * 0.95
            if def.low ? y0 < near : y1 > near {
                y0 = Swift.min(y0, over)
                y1 = Swift.max(y1, over)
            }
        }
        let ypad = Swift.max((y1 - y0) * 0.15, 1e-6)
        y0 -= ypad
        y1 += ypad

        func X(_ i: Int) -> CGFloat { pad + CGFloat(i) / CGFloat(n - 1) * (size.width - pad * 2) }
        func Y(_ v: Double) -> CGFloat {
            pad + CGFloat((y1 - v) / (y1 - y0)) * (size.height - pad * 2)
        }
        func clampY(_ v: Double) -> CGFloat { Swift.min(Swift.max(Y(v), pad), size.height - pad) }

        // The bands: past `over` in the danger tint, between the lines lighter.
        if let watch = def.watch, let over = def.over {
            func band(_ from: Double, _ to: Double, _ opacity: Double) {
                let ya = clampY(to), yb = clampY(from)
                guard abs(yb - ya) > 0.5 else { return }
                context.fill(
                    Path(CGRect(x: 0, y: Swift.min(ya, yb), width: size.width, height: abs(yb - ya))),
                    with: .color(Color(.danger).opacity(opacity))
                )
            }
            band(def.low ? y0 : over, def.low ? over : y1, 0.18)
            band(def.low ? over : watch, def.low ? watch : over, 0.09)
        }

        var line = Path()
        for (i, s) in series.enumerated() {
            let point = CGPoint(x: X(index[s.chIdx] ?? i), y: Y(s.v))
            if i == 0 { line.move(to: point) } else { line.addLine(to: point) }
        }
        context.stroke(line, with: .color(Color(.chartLine)), lineWidth: 1.5)

        // The highlighted laps, in the slot colours the chips use.
        for (slot, chIdx) in lit.enumerated() where slot < slots.count {
            guard let s = series.first(where: { $0.chIdx == chIdx }), let i = index[chIdx] else { continue }
            let c = CGPoint(x: X(i), y: Y(s.v))
            context.fill(
                Path(ellipseIn: CGRect(x: c.x - 2.5, y: c.y - 2.5, width: 5, height: 5)),
                with: .color(slots[slot])
            )
        }
    }

    // MARK: - Spread and fuel

    /// Cross-corner spread is the figure a setup change is judged by, so it gets
    /// words rather than a chart: left−right on each axle and front−rear, for
    /// the last lap that carried all four corners.
    @ViewBuilder
    private func spreadLines() -> some View {
        let temps = Health.sessionSpread(channels, "tyreC").last
        let pressures = Health.sessionSpread(channels, "tyreKpa").last
        if temps != nil || pressures != nil {
            VStack(alignment: .leading, spacing: 4) {
                Text("Cross-corner spread")
                    .teStyle(.eyebrow)
                    .foregroundStyle(Color(.textFaint))
                if let temps, let def = Health.defFor("tyreCLF") {
                    Text(spreadText(def, temps, "Tyre temps"))
                        .teStyle(.xs)
                        .foregroundStyle(Color(.textMuted))
                }
                if let pressures, let def = Health.defFor("tyreKpaLF") {
                    Text(spreadText(def, pressures, "Tyre pressures"))
                        .teStyle(.xs)
                        .foregroundStyle(Color(.textMuted))
                }
            }
        }
    }

    private func spreadText(_ def: Health.Def, _ spread: Health.LapSpread, _ label: String) -> String {
        let front = Health.displayDelta(def, spread.front, Self.units).text
        let rear = Health.displayDelta(def, spread.rear, Self.units).text
        let axle = Health.displayDelta(def, spread.axle, Self.units).text
        return "\(label) — front L−R \(front), rear L−R \(rear), front−rear \(axle)"
    }

    @ViewBuilder
    private func fuelLine() -> some View {
        if let fuel = Health.fuelBurn(channels) {
            Text(
                "Fuel — \(Int(fuel.perLapPct.rounded()))% a lap, \(Int(fuel.lastPct.rounded()))% left: "
                    + "≈\(fuel.lapsRemaining) lap\(fuel.lapsRemaining == 1 ? "" : "s") at this rate"
            )
            .teStyle(.xs)
            .foregroundStyle(Color(.textMuted))
        }
    }

    /// What VoiceOver gets: the stats line's own sentence when there is one,
    /// else every figure by name — a grid of sparklines is exactly what a
    /// screen-reader user cannot see.
    private func summary(_ sh: Health.SessionHealth) -> String {
        if let line = Health.healthSummary(channels, Self.units) { return line }
        return sh.columns.compactMap { column -> String? in
            guard let def = Health.defFor(column.key) else { return nil }
            return "\(def.label) \(Health.displayValue(def, column.extreme.v, Self.units).text)"
        }
        .joined(separator: ", ")
    }
}
